extensions/interactions/NumberWithUnits/directives/NumberWithUnitsValidationServiceSpec.js

```diff
@@ -0,0 +1,205 @@
+      },
+      labelled_as_correct: false,
+      param_changes: [],
+      refresher_exploration_id: null
+    });
+
+    equalsTwoRule = rof.createFromBackendDict({
+      rule_type: 'IsEqualTo',
+      inputs: {
+        f: createNumberWithUnitsDict('real', 2, createFractionDict(
+          false, 0, 0, 1), createUnitsDict({kg: 1, m: -2}))
+      }
+    });
+
+    equivalentToTwoThousandRule = rof.createFromBackendDict({
+      rule_type: 'IsEquivalentTo',
+      inputs: {
+        f: createNumberWithUnitsDict('real', 2000, createFractionDict(
+          false, 0, 0, 1), createUnitsDict({g: 1, m: -2}))
+      }
+    });
+
+    equivalentToTwoRule = rof.createFromBackendDict({
+      rule_type: 'IsEquivalentTo',
+      inputs: {
+        f: createNumberWithUnitsDict('real', 2, createFractionDict(
+          false, 0, 0, 1), createUnitsDict({kg: 1, m: -2}))
+      }
+    });
+
+    equalsTwoByThreeRule = rof.createFromBackendDict({
+      rule_type: 'IsEqualTo',
+      inputs: {
+        f: createNumberWithUnitsDict('fraction', 0, createFractionDict(
+          false, 0, 2, 3), createUnitsDict({kg: 1, m: -2}))
+      }
+    });
+
+    equivalentToTwoByThreeRule = rof.createFromBackendDict({
+      rule_type: 'IsEquivalentTo',
+      inputs: {
+        f: createNumberWithUnitsDict('fraction', 0, createFractionDict(
+          false, 0, 2000, 3), createUnitsDict({g: 1, m: -2}))
+      }
+    });
+
+    answerGroups = [agof.createNew(
+      [equalsTwoRule, equalsTwoByThreeRule],
+      goodDefaultOutcome,
+      false
+    )];
+  }));
+
+  it('should be able to perform basic validation', function() {
+    var warnings = validatorService.getAllWarnings(
+      currentState, {}, answerGroups, goodDefaultOutcome);
+    expect(warnings).toEqual([]);
+  });
+
+  it('should catch equals followed by equals same value', function() {
+    answerGroups[0].rules = [equalsTwoRule, equalsTwoRule];
+    var warnings = validatorService.getAllWarnings(
+      currentState, {}, answerGroups, goodDefaultOutcome);
+    expect(warnings).toEqual([{
+      type: WARNING_TYPES.ERROR,
+      message: 'Rule 2 from answer group 1 will never be matched ' +
+        'because it is made redundant by rule 1 from answer group 1.'
+    }]);
+  });
+
+  it('should not catch equals followed by equals with unequal values',
+    function() {
+      answerGroups[0].rules = [equalsTwoRule, equalsTwoByThreeRule];
+      var warnings = validatorService.getAllWarnings(
+        currentState, {}, answerGroups, goodDefaultOutcome);
+      expect(warnings).toEqual([]);
+    });
+
+  it('should not catch equals followed by equivalent with unequal values',
+    function() {
+      answerGroups[0].rules = [equalsTwoRule, equivalentToTwoThousandRule];
+      var warnings = validatorService.getAllWarnings(
+        currentState, {}, answerGroups, goodDefaultOutcome);
+      expect(warnings).toEqual([]);
+    });
+
+  it('should catch equals followed by equivalent with same value', function() {
```

Comment:
Rule 2 from answer group 1 can be matched if the answer is not exactly equal to but equivalent to given value. I've described this above, too.

Reply:
Ahh.. Ok. This is the flaw here. I forget to consider this case.

Reply:
Done, it's now allowed.